src/main/java/emu/grasscutter/command/source/impl/ClientChatCommandSource.java

```diff
@@ -0,0 +1,79 @@
+package emu.grasscutter.command.source.impl;
+
+import emu.grasscutter.command.handler.ContextFields;
+import emu.grasscutter.command.handler.HandlerContext;
+import emu.grasscutter.command.parser.annotation.Origin;
+import emu.grasscutter.command.source.BaseCommandSource;
+import emu.grasscutter.game.player.Player;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+import static emu.grasscutter.command.parser.commands.TargetCommand.PERSISTED_TARGET_KEY;
+
+public class ClientChatCommandSource extends BaseCommandSource {
+    private final Player player;
+    private boolean disposed = false;
+
+    private ClientChatCommandSource(Player player) {
+        this.player = player;
+    }
+    @Override
+    public String[] getPermissions() {
+        return player.getAccount().getPermissions().toArray(new String[] {});
+    }
+
+    @Override
+    public Origin getOrigin() {
+        return Origin.CLIENT;
+    }
+
+    @Override
+    public synchronized void info(String message) {
+        if (!disposed) {
+            player.dropMessage(message);
+        }
+    }
+
+    @Override
+    public synchronized void error(String error) {
+        if (!disposed) {
+            player.dropMessage("Error: %s".formatted(error));
+        }
+    }
+
+    @Override
+    public HandlerContext buildContext(HandlerContext.HandlerContextBuilder contextBuilder) {
+        HandlerContext context = contextBuilder
+                .errorConsumer(e -> error(e.toString()))
+                .resultConsumer(r -> info(r != null ? "Handler result: %s.".formatted(r.toString()) : "Handler completed."))
+                .messageConsumer(m -> info(m.toString()))
+                .build();
+        if (context.getContent().get(ContextFields.TARGET_UID) == null) {
+            Integer targetUid = (Integer) get(PERSISTED_TARGET_KEY);
+            if (targetUid != null) {
```

Comment:
Does this default to the in-game sender elsewhere?

Reply:
In both `ChatManager#sendPrivateMessage` and `ChatManager#sendPrivateMessage`

Reply:
Currently target priority works like this:

1. @UID as an argument inside command
2. Recipient of message in in-game chat
3. Persistent target set by `/target`
4. Sender

> In both `ChatManager#sendPrivateMessage` and `ChatManager#sendPrivateMessage`

That just calls `ClientChatCommandSource.getCommandSourceByPlayer(player).runCommand(message.substring(1));`, how does it handle 2. or 4.?

Reply:
4. can be handled by overriding `ClientChatSource#buildContext`.
2. can be handled by override `runCommand`. add a parameter `recipient`, save it to `persistedContext`, then call `super.runCommand`

Reply:
@uid as an argument cannot be parsed now. you call tell the program how to deal with it in `ParseUtil.PermittedClasses`

Reply:
> '4.' can be handled by overriding `ClientChatSource#buildContext`. '2.' can be handled by override `runCommand`. add a parameter `recipient`, save it to `persistedContext`, then call `super.runCommand`

> @uid as an argument cannot be parsed now. you call tell the program how to deal with it in `ParseUtil.PermittedClasses`

Ok, can you add those to your PR?

Reply:
> 1. @uid as an argument inside command

Done. and custom argument types are now supported.